### ISSUE TITLE
Exit status of `ansible-playbook` not as documented

### ISSUE THREAD

**Issue by kolibri** [ansible/ansible#19720]
##### ISSUE TYPE
 - Docs Report
- ~~Bug Report~~

##### COMPONENT NAME
`ansible-playbook` command.

##### ANSIBLE VERSION
```
 > ansible --version
ansible 2.3.0 (devel a10209cbe9) last updated 2016/12/28 22:13:02 (GMT +200)
  config file = /Users/ko/Development/projects/new-nest/ansible.cfg
  configured module search path = Default w/o overrides

 > ansible-playbook --version
ansible-playbook 2.3.0 (devel a10209cbe9) last updated 2016/12/28 22:13:02 (GMT +200)
  config file = /Users/ko/Development/projects/new-nest/ansible.cfg
  configured module search path = Default w/o overrides
```

##### CONFIGURATION
```
[defaults]
display_skipped_hosts = False
retry_files_enabled = False
roles_path = ./roles:./galaxy-roles
```
(but should not be related to the problem)

##### OS / ENVIRONMENT
Running Ansible from macOS `10.12.1 (16B2555)`

##### SUMMARY
The exit codes from the `ansible-playbook` command with an unreachable host is not as reported in `man ansible-playbook` (`4` instead of `3`).

##### STEPS TO REPRODUCE
(Just follow the code)
```yaml
# playbook.yml
- hosts: all
  tasks:
    - debug:  var=hostvars
```

```
# inventory
foobar.baz # Be shure, this host does not exist o_O
```

```bash
# then run
ansible-playbook -i inventory playbook.yml
echo "Exit code is: $?"
```

This as a gist: https://gist.github.com/kolibri/744effd6d713b69928aa9b9bddea291a
(clone an `./run.sh`)

##### EXPECTED RESULTS
Exit code `3` when a host ist unreachable (or, if this is not an error, a more understandable documentation what `unreachable` means) 

##### ACTUAL RESULTS
```
Using /etc/ansible/ansible.cfg as config file

PLAYBOOK: playbook.yml ***************************************************************************************************************************
1 plays in playbook.yml

PLAY [all] ***************************************************************************************************************************************

TASK [Gathering Facts] ***************************************************************************************************************************
Using module file /Users/ko/Development/vendor/ansible/lib/ansible/modules/system/setup.py
<foobar.baz> ESTABLISH SSH CONNECTION FOR USER: None
<foobar.baz> SSH: EXEC ssh -C -o ControlMaster=auto -o ControlPersist=60s -o KbdInteractiveAuthentication=no -o PreferredAuthentications=gssapi-with-mic,gssapi-keyex,hostbased,publickey -o PasswordAuthentication=no -o ConnectTimeout=10 -o ControlPath=/Users/ko/.ansible/cp/ansible-ssh-%h-%p-%r foobar.baz '/bin/sh -c '"'"'( umask 77 && mkdir -p "` echo ~/.ansible/tmp/ansible-tmp-1482961771.21-8227272489736 `" && echo ansible-tmp-1482961771.21-8227272489736="` echo ~/.ansible/tmp/ansible-tmp-1482961771.21-8227272489736 `" ) && sleep 0'"'"''
fatal: [foobar.baz]: UNREACHABLE! => {
    "changed": false,
    "msg": "Failed to connect to the host via ssh: ssh: Could not resolve hostname foobar.baz: nodename nor servname provided, or not known\r\n",
    "unreachable": true
}

PLAY RECAP ***************************************************************************************************************************************
foobar.baz                 : ok=0    changed=0    unreachable=1    failed=0

Exit code is: 4
```
(this is not really necessary for this issue, is it? So, here a snippet from my `man ansible-playbook` that is more related: )
```
EXIT STATUS -----------

*0* -- OK or no hosts matched

*1* -- Error

*2* -- One or more hosts failed

*3* -- One or more hosts were unreachable

*4* -- Parser error

*5* -- Bad or incomplete options

*99* -- User interrupted execution

*250* -- Unexpected error
```

**Comment by sivel** [comment]
It looks like there is an inconsistency in the exit codes.  It appears as though `unreachable` was re-defined as `4`.

https://github.com/ansible/ansible/blob/devel/lib/ansible/executor/task_queue_manager.py#L61-L66

I'm not sure of the reason as of now, likely due to needs of bitwise operations: https://github.com/ansible/ansible/commit/c669a381

It also looks like there is inconsistency between 250/255.

**Comment by bcoca** [comment]
@sivel those are tqm codes, they don't map directly to Ansible exit codes

**Comment by bcoca** [comment]
@kolibri, 2 should read 'play failed due to failed and/or unreachable hosts'

**Comment by sivel** [comment]
@bcoca That is incorrect.  I traced the code, and it looks like tqm codes can make it through to become an exit code.

In the case of an unreachable host, you get an exit status of `4`.

This is a result of `exit_code = cli.run()` in `bin/ansible`.  The return from `cli.run` is one of `result = self._tqm.run(play)`.

So depending on the return of the tqm, you get specific exit codes.

In `bin/ansible`:

```
# TQM takes care of these, but leaving comment to reserve the exit codes
```

**Comment by sivel** [comment]
This does also apply to a `255` too, there is the possibility that the linear strategy will return a `255`, which will result in a 255 exit code as well.

**Comment by bcoca** [comment]
@sivel, but not all of them are possible, only a subset are returned, that is why it is not a direct maping.

As for rc = 4, it will only return in 2 cases (1 seems like a bug):

1) (correct) play exits due to all hosts in batch failing
2) (incorrect) host is unreachable and this is the last task/host run for play

255 and 250 are both similar codes, but not the same, we need language to make this clearer, 255 can be triggered by the strategy OR uncaught exceptions inside or outside of play execution.

**Comment by AllenKll** [comment]
This still exists in version 2.4.0

**Comment by shenberg** [comment]
@bcoca erm 4 is actually mapped twice as a return code differently: tqm returns 4 on unreachable hosts (any unreachable hosts, _not_ all), while `bin/ansible` [returns 4 on `AnsibleParseError`](https://github.com/ansible/ansible/blob/f8e3cfe9e25d889f15aae4d779b12853aa50e9f7/bin/ansible#L115)

**Comment by ssbarnea** [comment]
Our of curiosity, is there a way to make ansible exist with a special (custom) error code in case something happened?

I want to use a special exist code for the case when some unittests failed and exist code is the best way to communicate with CI processes.

**Comment by matonb** [comment]
Any progress on sorting out parser error and host(s) unreachable both exiting status 4?

**Comment by gdubicki** [comment]
Please create a separate issue for that idea @ssbarnea . I would love to have that feature too, btw!

**Comment by Meera94** [comment]
What is the status of the fix? It would be useful to get the documented exit codes since my scripts can then rely on them.

**Comment by kladiv** [comment]
+1

**Comment by smothiki** [comment]
We are also facing this issue. If there are any unreachable hosts or ssh login failures present in the RUN ansible-playbook is returning 4 exit code

**Comment by acozine** [comment]
Tested today. The `ansible-playbook` command still returns 4 as an exit code when a host is unreachable. More generally, exit codes are inconsistent. We do not currently document this behavior (incorrectly, or at all) in any documentation we control. This means it's a low priority for now.

If you integrate Ansible with other systems, we recommend testing the return codes you need to use. Closing this issue - feel free to open a fresh one if this is an ongoing concern.

**Comment by gdubicki** [comment]
I don’t understand „We do not currently document this behavior (incorrectly, or at all) in any documentation we control” from your reply, @acozine . It is documented in man. Is man for ansible-playbook outside your control?

**Comment by acozine** [comment]
@gdubicki can you provide a current link, or an installation method and version, that would get me to the man page where you see this information? I've checked all source material we maintain, and we do not document it anywhere that I can find. It's possible that someone involved in packaging for a particular platform added this information to their man page.

**Comment by ssbarnea** [comment]
Thanks for reopening it. I know for sure that documenting that exit code 2 means some hosts failed (partial success) is key for anyone using ansible, as they can add extra conditions for treating this particular code as a success (if they expect a partial success).

**Comment by samccann** [comment]
@gundalow @ssbarnea as was mentioned in earlier comments, we have no control over man pages so cannot address that particular issue.

We can document exit codes in docs.ansible.com, though.

Taking a prior link - are these the codes we should document? (@sivel @bcoca ) https://github.com/ansible/ansible/blob/devel/lib/ansible/executor/task_queue_manager.py#L95

 RUN_OK = 0
    RUN_ERROR = 1
    RUN_FAILED_HOSTS = 2
    RUN_UNREACHABLE_HOSTS = 4
    RUN_FAILED_BREAK_PLAY = 8
    RUN_UNKNOWN_ERROR = 255


Or are they from here:
https://github.com/ansible/ansible/blob/devel/lib/ansible/cli/scripts/ansible_cli_stub.py#L134

TQM takes care of these, but leaving comment to reserve the exit codes
    except AnsibleHostUnreachable as e:
        display.error(str(e))
        exit_code = 3
    except AnsibleHostFailed as e:
        display.error(str(e))
        exit_code = 2

Finally, the cli pages in docs.ansible.com are generated from the code and there is no place I can see to add these error codes so they show up on the [cli docs page](https://docs.ansible.com/ansible/latest/cli/ansible-playbook.html#description) unless we add them to the description at https://github.com/ansible/ansible/blob/devel/lib/ansible/cli/playbook.py#L29
